Fix MSVC compile issue related with ssize_t

denix56 opened the pull request:
The issue is caused  because in MSVC the gnu ssize_t type is called SSIZE_T

Maintainer edit: Fixes #2023.

SergioRAgostinho commented:
I ran a quick find through cpl and this type is only used in the Dinast Grabber cpp file, while using a libusb function, and as expected, they export that type. See http://libusb.sourceforge.net/api-1.0/libusb_8h_source.html line 36 onwards. So what exactly is happening in your case?

denix56 commented:
msvc throws error while compiling pcd_io.cpp
look somewhere near line 770:

`ssize_t num_read = pcl_read (fd, &compressed_size, 4);`

hear is the use of ssize_t

SergioRAgostinho commented:
It was introduced on a fairly recent merge. So I would prefer to follow what both MingW and libusb do and copy the snippet below here https://github.com/PointCloudLibrary/pcl/blob/master/io/src/pcd_io.cpp#L59, on the Windows related conditional branch.
```cpp
/* ssize_t is also not available (copy/paste from MinGW) */
#ifndef _SSIZE_T_DEFINED
#define _SSIZE_T_DEFINED
#undef ssize_t
#ifdef _WIN64
typedef __int64 ssize_t;
#else
typedef int ssize_t;
#endif /* _WIN64 */
#endif /* _SSIZE_T_DEFINED */
#endif /* _MSC_VER */
```

UnaNancyOwen commented:
That line seems to be included in since release PCL 1.8.1. (PR #1986)
It is probably affects all versions of Visual C++.
I think that it is necessary resolve this problem.

SergioRAgostinho commented:
Scrap what I just said, copy the snippet here https://github.com/PointCloudLibrary/pcl/blob/master/io/include/pcl/io/impl/pcd_io.hpp#L63 . It has to go there because that's where we're exporting this `pcl_read` definition which is always accessible if you include the header.

UnaNancyOwen commented:
@SergioRAgostinho It can be built successfully. But, test failed.
(However, I don't know whether this test has failed since before on Windows.)
```
> .\test_grabbers.exe "../../test/grabber_sequences"
added: ../../test/grabber_sequences/pcd\frame_20121214T142255.814212.pcd
added: ../../test/grabber_sequences/pcd\frame_20121214T142256.068683.pcd
added: ../../test/grabber_sequences/pcd\frame_20121214T142256.332395.pcd
[==========] Running 7 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 7 tests from PCL
[ RUN      ] PCL.PCDGrabber
C:\pcl-FixMSVCTypeIssue\test\io\test_grabbers.cpp(56): error:       Expected: pcds_.size ()
      Which is: 3
To be equal to: grabbed_clouds.size ()
      Which is: 1
[  FAILED  ] PCL.PCDGrabber (3319 ms)
[ RUN      ] PCL.ImageGrabberTIFF
[       OK ] PCL.ImageGrabberTIFF (5562 ms)
[ RUN      ] PCL.ImageGrabberPCLZF
[       OK ] PCL.ImageGrabberPCLZF (5558 ms)
[ RUN      ] PCL.ImageGrabberOMP
User Error 1001: argument to num_threads clause must be positive
User Error 1001: argument to num_threads clause must be positive
User Error 1001: argument to num_threads clause must be positive
User Error 1001: argument to num_threads clause must be positive
User Error 1001: argument to num_threads clause must be positive
User Error 1001: argument to num_threads clause must be positive
[       OK ] PCL.ImageGrabberOMP (4798 ms)
[ RUN      ] PCL.ImageGrabberTimestamps
[       OK ] PCL.ImageGrabberTimestamps (8 ms)
[ RUN      ] PCL.ImageGrabberSetIntrinsicsTIFF
[       OK ] PCL.ImageGrabberSetIntrinsicsTIFF (2494 ms)
[ RUN      ] PCL.ImageGrabberSetIntrinsicsPCLZF
[       OK ] PCL.ImageGrabberSetIntrinsicsPCLZF (2413 ms)
[----------] 7 tests from PCL (24168 ms total)

[----------] Global test environment tear-down
[==========] 7 tests from 1 test case ran. (24169 ms total)
[  PASSED  ] 6 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] PCL.PCDGrabber

 1 FAILED TEST
```

denix56 commented:
It is strange. I ran that test too on Visual Studio 2017:

```
added: D:\dev-tools\pcl\test\grabber_sequences/pcd\frame_20121214T142255.814212.pcd
added: D:\dev-tools\pcl\test\grabber_sequences/pcd\frame_20121214T142256.068683.pcd
added: D:\dev-tools\pcl\test\grabber_sequences/pcd\frame_20121214T142256.332395.pcd
[==========] Running 7 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 7 tests from PCL
[ RUN      ] PCL.PCDGrabber
[       OK ] PCL.PCDGrabber (1339 ms)
[ RUN      ] PCL.ImageGrabberTIFF
[       OK ] PCL.ImageGrabberTIFF (540 ms)
[ RUN      ] PCL.ImageGrabberPCLZF
[       OK ] PCL.ImageGrabberPCLZF (355 ms)
[ RUN      ] PCL.ImageGrabberOMP
[       OK ] PCL.ImageGrabberOMP (314 ms)
[ RUN      ] PCL.ImageGrabberTimestamps
[       OK ] PCL.ImageGrabberTimestamps (0 ms)
[ RUN      ] PCL.ImageGrabberSetIntrinsicsTIFF
[       OK ] PCL.ImageGrabberSetIntrinsicsTIFF (221 ms)
[ RUN      ] PCL.ImageGrabberSetIntrinsicsPCLZF
[       OK ] PCL.ImageGrabberSetIntrinsicsPCLZF (173 ms)
[----------] 7 tests from PCL (2945 ms total)

[----------] Global test environment tear-down
[==========] 7 tests from 1 test case ran. (2946 ms total)
[  PASSED  ] 7 tests.
```

UnaNancyOwen commented:
@denix56 What do you use updated version of Visual Studio 2017? What is version numbers of x? (15.x)

UnaNancyOwen commented:
@denix56 @SergioRAgostinho Sorry, I tried test again. It all succeeded. I think ready to merge this pull request.
```
> .\test_grabbers.exe "../../test/grabber_sequences"
added: ../../test/grabber_sequences/pcd\frame_20121214T142255.814212.pcd
added: ../../test/grabber_sequences/pcd\frame_20121214T142256.068683.pcd
added: ../../test/grabber_sequences/pcd\frame_20121214T142256.332395.pcd
[==========] Running 7 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 7 tests from PCL
[ RUN      ] PCL.PCDGrabber
[       OK ] PCL.PCDGrabber (1293 ms)
[ RUN      ] PCL.ImageGrabberTIFF
[       OK ] PCL.ImageGrabberTIFF (649 ms)
[ RUN      ] PCL.ImageGrabberPCLZF
[       OK ] PCL.ImageGrabberPCLZF (258 ms)
[ RUN      ] PCL.ImageGrabberOMP
[       OK ] PCL.ImageGrabberOMP (1041 ms)
[ RUN      ] PCL.ImageGrabberTimestamps
[       OK ] PCL.ImageGrabberTimestamps (1 ms)
[ RUN      ] PCL.ImageGrabberSetIntrinsicsTIFF
[       OK ] PCL.ImageGrabberSetIntrinsicsTIFF (182 ms)
[ RUN      ] PCL.ImageGrabberSetIntrinsicsPCLZF
[       OK ] PCL.ImageGrabberSetIntrinsicsPCLZF (163 ms)
[----------] 7 tests from PCL (3593 ms total)

[----------] Global test environment tear-down
[==========] 7 tests from 1 test case ran. (3595 ms total)
[  PASSED  ] 7 tests.
```
